controllers/cluster_controller_phases.go

```diff
@@ -178,24 +178,63 @@ func (r *ClusterReconciler) reconcileInfrastructure(ctx context.Context, cluster
 	return nil
 }
 
+// reconcileControlPlane reconciles the Spec.ControlPlaneRef object on a Cluster.
+func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, cluster *clusterv1.Cluster) error {
+	if cluster.Spec.ControlPlaneRef == nil {
+		return nil
+	}
+
+	// Call generic external reconciler.
+	controlPlaneConfig, err := r.reconcileExternal(ctx, cluster, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return err
+	}
+
+	// There's no need to go any further if the control plane resource is marked for deletion.
+	if !controlPlaneConfig.GetDeletionTimestamp().IsZero() {
+		return nil
+	}
+
+	// Update cluster.Status.ControlPlaneInitialized if it hasn't already been set
+	if !cluster.Status.ControlPlaneInitialized {
+		if err := util.UnstructuredUnmarshalField(controlPlaneConfig, cluster.Status.ControlPlaneInitialized, "status", "initialized"); err != nil {
+			return errors.Wrapf(err, "failed to retrieve Status.ControlPlaneInitialized from control plane provider for Cluster %q in namespace %q",
+				cluster.Name, cluster.Namespace)
+		}
+	}
+
+	// Determine if the control plane provider is ready.
+	ready, err := external.IsReady(controlPlaneConfig)
+	if err != nil {
+		return err
+	}
+	cluster.Status.ControlPlaneReady = ready
+
+	return nil
+}
+
 func (r *ClusterReconciler) reconcileKubeconfig(ctx context.Context, cluster *clusterv1.Cluster) error {
 	if cluster.Spec.ControlPlaneEndpoint.IsZero() {
 		return nil
 	}
 
-	_, err := secret.Get(r.Client, cluster, secret.Kubeconfig)
-	switch {
-	case apierrors.IsNotFound(err):
-		if err := kubeconfig.CreateSecret(ctx, r.Client, cluster); err != nil {
-			if err == kubeconfig.ErrDependentCertificateNotFound {
-				return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second},
-					"could not find secret %q for Cluster %q in namespace %q, requeuing",
-					secret.ClusterCA, cluster.Name, cluster.Namespace)
+	// Only generate the Kubeconfig if there is no ControlPlaneRef
+	if cluster.Spec.ControlPlaneRef == nil {
```

Comment:
If you want to save on nesting, you can invert this to a `!= nil` check and return early